models/jupiter/solana/jupiter_solana_perp_events.sql

```diff
@@ -0,0 +1,113 @@
+ {{
+  config(
+        schema = 'jupiter_solana',
+        alias = 'perp_events',
+        partition_by = ['block_month'],
+        materialized='incremental',
+        file_format = 'delta',
+        incremental_strategy='merge',
+        unique_key = ['position_change','position_key','tx_id'],
+        post_hook='{{ expose_spells(\'["jupiter"]\',
+                                    "project",
+                                    "jupiter_solana",
+                                    \'["ilemi"]\') }}')
+}}
+
+-- IncreasePositionEvent
+    --create position request
+    --then fillers spam create increase/decrease until it settles (preswap can happen but just ignore it)
+    --orders are aggregated under same position key
+    --example long https://solana.fm/tx/3to8mJ9DPmb3vSXXLF7QU2vjggHW21JoGCwtuJ29HKVvbf8iBWHetFmf3Ewx8E54XA2ypvQkqZqKajFJpYx8Qxjd?cluster=mainnet-qn1
+    --example short https://solana.fm/tx/5eUVyRYyFVUX7vBZLhqbz3ahLTYLPzYSfpFskdukAvtQatEAT1adEzvrdWn7xoAwuL5wYjMFG2o7PtXGFjoT5fKa?cluster=mainnet-qn1
+SELECT
+    'increase' AS position_change
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+291,8))) / 1e6 as size_usd --this is the levered size delta (change)
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+299,8))) / 1e6 as collateral_usd --collateral size, transferred into pool
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+307,8))) as collateral_token --collateral notional size, transferred into pool
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+340,8))) / 1e6 as fee_usd --open fee
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+315,8))) / 1e6 as price_usd --entry price
+    , null as liq_fee_usd
+    , null as pnl_direction
+    , null as pnl_usd --no pnl on increase
+    , to_base58(bytearray_substring(data,1+227,32)) as owner
+    , to_base58(bytearray_substring(data,1+16,32)) as position_key
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+16+32,1))) as position_side --1 is long, 2 is short
+    , to_base58(bytearray_substring(data,1+16+32+1,32)) as custody_position_key --ties to the token mint of long or short token
+    , to_base58(bytearray_substring(data,1+16+32+1+32,32)) as custody_collateral_key --ties to the token mint of the collateral token
+    , block_slot
+    , block_time
+    , cast(date_trunc('month', block_time) as date) as block_month
```

Comment:
i'm noticing `block_time` can be null in this source table, is that correct?
if yes, then we may need to reconsider as a partition column, as we can't write nulls

Reply:
it is null only in 2021. Does not affect perps which were deployed in 2023